File: crossecore-java/src/java_/AbstractMethodDeclarationBase.java
package java_;
import org.eclipse.emf.common.notify.Notification;
import org.eclipse.emf.common.notify.NotificationChain;
import org.eclipse.emf.common.notify.impl.NotificationImpl;
import org.eclipse.emf.ecore.InternalEObject;
import org.eclipse.emf.ecore.*;
import org.eclipse.emf.ecore.impl.ENotificationImpl;
public class AbstractMethodDeclarationBase 
extends BodyDeclarationImpl implements AbstractMethodDeclaration
{
	private Block _body;
	public Block getBody()
	{
		
			return _body;
		
		
	}
	public void setBody(Block value){
		if (value != _body) {
			NotificationChain msgs = null;
			if (_body != null){
				msgs = ((InternalEObject)_body).eInverseRemove(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__BODY, null, msgs);
			}
			if (value != null){
				msgs = ((InternalEObject)value).eInverseAdd(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__BODY, null, msgs);
			}
			msgs = basicSetBody(value, msgs);
			if (msgs != null) {
				msgs.dispatch();
			}
		}
		else if (eNotificationRequired()){
			eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__BODY , value, value));
		}
	}
	private com.crossecore.ocl.OrderedSet<SingleVariableDeclaration> _parameters;
	
	public com.crossecore.ocl.OrderedSet<SingleVariableDeclaration> getParameters()
	{
		if(_parameters==null){
			_parameters = new com.crossecore.ocl.OrderedSet<SingleVariableDeclaration>(SingleVariableDeclaration.class, this, Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__PARAMETERS, Java_PackageImpl.SINGLE_VARIABLE_DECLARATION__METHOD_DECLARATION);
		}
		return _parameters;
	
	}
	private com.crossecore.ocl.OrderedSet<TypeAccess> _thrownExceptions;
	
	public com.crossecore.ocl.OrderedSet<TypeAccess> getThrownExceptions()
	{
		if(_thrownExceptions==null){
			_thrownExceptions = new com.crossecore.ocl.OrderedSet<TypeAccess>(TypeAccess.class, this, Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__THROWN_EXCEPTIONS, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__THROWN_EXCEPTIONS);
		}
		return _thrownExceptions;
	
	}
	private com.crossecore.ocl.OrderedSet<TypeParameter> _typeParameters;
	
	public com.crossecore.ocl.OrderedSet<TypeParameter> getTypeParameters()
	{
		if(_typeParameters==null){
			_typeParameters = new com.crossecore.ocl.OrderedSet<TypeParameter>(TypeParameter.class, this, Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__TYPE_PARAMETERS, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__TYPE_PARAMETERS);
		}
		return _typeParameters;
	
	}
	private com.crossecore.ocl.Set<MethodRef> _usagesInDocComments;
	
	public com.crossecore.ocl.Set<MethodRef> getUsagesInDocComments()
	{
		if(_usagesInDocComments==null){
			_usagesInDocComments = new com.crossecore.ocl.Set<MethodRef>(MethodRef.class, this, Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES_IN_DOC_COMMENTS, Java_PackageImpl.METHOD_REF__METHOD);
		}
		return _usagesInDocComments;
	
	}
	private com.crossecore.ocl.Set<AbstractMethodInvocation> _usages;
	
	public com.crossecore.ocl.Set<AbstractMethodInvocation> getUsages()
	{
		if(_usages==null){
			_usages = new com.crossecore.ocl.Set<AbstractMethodInvocation>(AbstractMethodInvocation.class, this, Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES, Java_PackageImpl.ABSTRACT_METHOD_INVOCATION__METHOD);
		}
		return _usages;
	
	}

	
	@Override
	protected EClass eStaticClass() {
		return Java_PackageImpl.Literals.ABSTRACTMETHODDECLARATION;
	}
	
	@Override
	public NotificationChain eInverseAdd(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES_IN_DOC_COMMENTS:
				return getUsagesInDocComments().basicAdd((MethodRef)otherEnd, msgs);
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES:
				return getUsages().basicAdd((AbstractMethodInvocation)otherEnd, msgs);
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__PARAMETERS:
				return getParameters().basicAdd((SingleVariableDeclaration)otherEnd, msgs);
		}
		return super.eInverseAdd(otherEnd, featureID, msgs);
	}
	
	@Override
	public NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES_IN_DOC_COMMENTS:
				return getUsagesInDocComments().basicRemove((MethodRef)otherEnd, msgs);
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES:
				return getUsages().basicRemove((AbstractMethodInvocation)otherEnd, msgs);
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__PARAMETERS:
				return getParameters().basicRemove((SingleVariableDeclaration)otherEnd, msgs);
		}
		return super.eInverseRemove(otherEnd, featureID, msgs);
	}
	
	public NotificationChain basicSetBody(Block newobj, NotificationChain msgs) {
		Block oldobj = _body;
		_body = newobj;
		if (eNotificationRequired()) {
			ENotificationImpl notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__BODY, oldobj, newobj);
			if (msgs == null){
				msgs = notification;
			}
			else{
				msgs.add(notification);
			}
		}
		return msgs;
	}
	
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__BODY:
				return getBody();
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__PARAMETERS:
				return getParameters();
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__THROWN_EXCEPTIONS:
				return getThrownExceptions();
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__TYPE_PARAMETERS:
				return getTypeParameters();
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES_IN_DOC_COMMENTS:
				return getUsagesInDocComments();
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES:
				return getUsages();
		}
		return super.eGet(featureID, resolve, coreType);
	}
	
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__BODY:
				setBody((Block) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__PARAMETERS:
				getParameters().clear();
				getParameters().addAll((java.util.Collection<? extends SingleVariableDeclaration>) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__THROWN_EXCEPTIONS:
				getThrownExceptions().clear();
				getThrownExceptions().addAll((java.util.Collection<? extends TypeAccess>) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__TYPE_PARAMETERS:
				getTypeParameters().clear();
				getTypeParameters().addAll((java.util.Collection<? extends TypeParameter>) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES_IN_DOC_COMMENTS:
				getUsagesInDocComments().clear();
				getUsagesInDocComments().addAll((java.util.Collection<? extends MethodRef>) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES:
				getUsages().clear();
				getUsages().addAll((java.util.Collection<? extends AbstractMethodInvocation>) newValue);
				return;
		}
		super.eSet(featureID, newValue);
	}
	
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__BODY:
				return _body != null; //single != null;
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__PARAMETERS:
				return _parameters != null && !_parameters.isEmpty();
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__THROWN_EXCEPTIONS:
				return _thrownExceptions != null && !_thrownExceptions.isEmpty();
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__TYPE_PARAMETERS:
				return _typeParameters != null && !_typeParameters.isEmpty();
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES_IN_DOC_COMMENTS:
				return getUsagesInDocComments().isEmpty(); //many, volatile
			case Java_PackageImpl.ABSTRACT_METHOD_DECLARATION__USAGES:
				return getUsages().isEmpty(); //many, volatile
		}
		return super.eIsSet(featureID);
	}
	
	
	
}

File: crossecore-csharp/Java/Java/Java_/AbstractMethodDeclarationBase.cs
/* CrossEcore is a cross-platform modeling framework that generates C#, TypeScript, 
 * JavaScript, Swift code from Ecore models with embedded OCL (http://www.crossecore.org/).
 * The original Eclipse Modeling Framework is available at https://www.eclipse.org/modeling/emf/.
 * 
 * contributor: Simon Schwichtenberg
 */
 
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using oclstdlib;
using Ecore;
namespace Java_{
	public class AbstractMethodDeclarationBase 
	:BodyDeclarationImpl, AbstractMethodDeclaration
	{
		private Block _body;
		public virtual Block body
		{
			get {
			
				return _body;
			}
			set {
				if (value != _body) {
					NotificationChain msgs = null;
					if (_body != null){
						msgs = ((InternalEObject)_body).eInverseRemove(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACTMETHODDECLARATION_BODY, null, msgs);
					}
					if (value != null){
						msgs = ((InternalEObject)value).eInverseAdd(this, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACTMETHODDECLARATION_BODY, null, msgs);
					}
					msgs = basicSetBody(value, msgs);
					if (msgs != null) {
						msgs.dispatch();
					}
				}
				else if (eNotificationRequired()){
					eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.ABSTRACTMETHODDECLARATION_BODY , value, value));
				}
				}
		}
		private OrderedSet<SingleVariableDeclaration> _parameters;
		
		public virtual OrderedSet<SingleVariableDeclaration> parameters
		{
			get {
				if(_parameters==null){
					_parameters = new OrderedSet<SingleVariableDeclaration>(this, Java_PackageImpl.ABSTRACTMETHODDECLARATION_PARAMETERS, Java_PackageImpl.SINGLEVARIABLEDECLARATION_METHODDECLARATION);
				}
				return _parameters;
			}
		
		}
		private OrderedSet<TypeAccess> _thrownExceptions;
		
		public virtual OrderedSet<TypeAccess> thrownExceptions
		{
			get {
				if(_thrownExceptions==null){
					_thrownExceptions = new OrderedSet<TypeAccess>(this, Java_PackageImpl.ABSTRACTMETHODDECLARATION_THROWNEXCEPTIONS, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACTMETHODDECLARATION_THROWNEXCEPTIONS);
				}
				return _thrownExceptions;
			}
		
		}
		private OrderedSet<TypeParameter> _typeParameters;
		
		public virtual OrderedSet<TypeParameter> typeParameters
		{
			get {
				if(_typeParameters==null){
					_typeParameters = new OrderedSet<TypeParameter>(this, Java_PackageImpl.ABSTRACTMETHODDECLARATION_TYPEPARAMETERS, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACTMETHODDECLARATION_TYPEPARAMETERS);
				}
				return _typeParameters;
			}
		
		}
		private Set<MethodRef> _usagesInDocComments;
		
		public virtual Set<MethodRef> usagesInDocComments
		{
			get {
				if(_usagesInDocComments==null){
					_usagesInDocComments = new Set<MethodRef>(this, Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGESINDOCCOMMENTS, Java_PackageImpl.METHODREF_METHOD);
				}
				return _usagesInDocComments;
			}
		
		}
		private Set<AbstractMethodInvocation> _usages;
		
		public virtual Set<AbstractMethodInvocation> usages
		{
			get {
				if(_usages==null){
					_usages = new Set<AbstractMethodInvocation>(this, Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGES, Java_PackageImpl.ABSTRACTMETHODINVOCATION_METHOD);
				}
				return _usages;
			}
		
		}
		
		protected override EClass eStaticClass() {
			return Java_PackageImpl.Literals.ABSTRACTMETHODDECLARATION;
		}
		
		public override NotificationChain eInverseAdd(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_PARAMETERS:
					return parameters.basicAdd((SingleVariableDeclaration)otherEnd, msgs);
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGESINDOCCOMMENTS:
					return usagesInDocComments.basicAdd((MethodRef)otherEnd, msgs);
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGES:
					return usages.basicAdd((AbstractMethodInvocation)otherEnd, msgs);
			}
			return base.eInverseAdd(otherEnd, featureID, msgs);
		}
		
		public override NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_PARAMETERS:
					return parameters.basicRemove((SingleVariableDeclaration)otherEnd, msgs);
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGESINDOCCOMMENTS:
					return usagesInDocComments.basicRemove((MethodRef)otherEnd, msgs);
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGES:
					return usages.basicRemove((AbstractMethodInvocation)otherEnd, msgs);
			}
			return base.eInverseRemove(otherEnd, featureID, msgs);
		}
		
		public NotificationChain basicSetBody(Block newobj, NotificationChain msgs) {
			var oldobj = _body;
			_body = newobj;
			if (eNotificationRequired()) {
				var notification = new ENotificationImpl(this, NotificationImpl.SET, Java_PackageImpl.ABSTRACTMETHODDECLARATION_BODY, oldobj, newobj);
				if (msgs == null){
					msgs = notification;
				}
				else{
					msgs.add(notification);
				}
			}
			return msgs;
		}
		
		public override object eGet(int featureID, bool resolve, bool coreType) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_BODY:
					return body;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_PARAMETERS:
					return parameters;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_THROWNEXCEPTIONS:
					return thrownExceptions;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_TYPEPARAMETERS:
					return typeParameters;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGESINDOCCOMMENTS:
					return usagesInDocComments;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGES:
					return usages;
			}
			return base.eGet(featureID, resolve, coreType);
		}
		
		
		public override void eSet(int featureID, object newValue) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_BODY:
					body = (Block) newValue;
					return;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_PARAMETERS:
					parameters.Clear();
					parameters.AddRange(((List<EObject>)newValue)?.Cast<SingleVariableDeclaration>());
					return;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_THROWNEXCEPTIONS:
					thrownExceptions.Clear();
					thrownExceptions.AddRange(((List<EObject>)newValue)?.Cast<TypeAccess>());
					return;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_TYPEPARAMETERS:
					typeParameters.Clear();
					typeParameters.AddRange(((List<EObject>)newValue)?.Cast<TypeParameter>());
					return;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGESINDOCCOMMENTS:
					usagesInDocComments.Clear();
					usagesInDocComments.AddRange(((List<EObject>)newValue)?.Cast<MethodRef>());
					return;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGES:
					usages.Clear();
					usages.AddRange(((List<EObject>)newValue)?.Cast<AbstractMethodInvocation>());
					return;
			}
			base.eSet(featureID, newValue);
		}
		
		/*
		public override bool eIsSet(int featureID) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_BODY:
					return _body != null; //single != null;
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_PARAMETERS:
					return _parameters != null && !_parameters.isEmpty();
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_THROWNEXCEPTIONS:
					return _thrownExceptions != null && !_thrownExceptions.isEmpty();
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_TYPEPARAMETERS:
					return _typeParameters != null && !_typeParameters.isEmpty();
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGESINDOCCOMMENTS:
					return getUsagesInDocComments().isEmpty(); //many, volatile
				case Java_PackageImpl.ABSTRACTMETHODDECLARATION_USAGES:
					return getUsages().isEmpty(); //many, volatile
			}
			return base.eIsSet(featureID);
		}
		*/
		
		
		
	}
}
